distributed/semaphore.py

```diff
@@ -0,0 +1,327 @@
+    * semaphore_release
+    * semaphore_create
+    """
+
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        self.leases = defaultdict(deque)
+        self.events = defaultdict(asyncio.Event)
+        self.max_leases = dict()
+        self.leases_per_client = defaultdict(partial(defaultdict, deque))
+        self.scheduler.handlers.update(
+            {
+                "semaphore_create": self.create,
+                "semaphore_acquire": self.acquire,
+                "semaphore_release": self.release,
+                "semaphore_close": self.close,
+            }
+        )
+
+        self.scheduler.extensions["semaphores"] = self
+        self.pc_validate_leases = PeriodicCallback(
+            self._validate_leases,
+            1000
+            * parse_timedelta(
+                dask.config.get(
+                    "distributed.scheduler.locks.lease-validation-interval"
+                ),
+                default="s",
+            ),
+            io_loop=self.scheduler.loop,
+        )
+        self.pc_validate_leases.start()
+        self._validation_running = False
+
+    # `comm` here is required by the handler interface
+    async def create(self, comm=None, name=None, max_leases=None):
+        # We use `self.max_leases.keys()` as the point of truth to find out if a semaphore with a specific
+        # `name` has been created.
+        if name not in self.max_leases:
+            assert isinstance(max_leases, int), max_leases
+            self.max_leases[name] = max_leases
+        else:
+            if max_leases != self.max_leases[name]:
+                raise ValueError(
+                    "Inconsistent max leases: %s, expected: %s"
+                    % (max_leases, self.max_leases[name])
+                )
+
+    async def _get_lease(self, client, name, identifier):
+        # We should make sure that the client is already properly registered with the scheduler
+        # otherwise the lease validation will mop up every acquired lease immediately. That's mostly relevant for tests
+        while client not in self.scheduler.clients:
+            await asyncio.sleep(0.0005)  # This value is set somewhat arbitrarily
+
+        result = True
+        if len(self.leases[name]) < self.max_leases[name]:
+            # naive: self.leases[resource] += 1
+            # not naive:
+            self.leases[name].append(identifier)
+            self.leases_per_client[client][name].append(identifier)
+        else:
+            result = False
+        return result
+
+    def _semaphore_exists(self, name):
+        if name not in self.max_leases:
+            logger.exception(f"Semaphore `{name}` not known or already closed.")
+            return False
+        return True
+
+    async def acquire(
+        self, comm=None, name=None, client=None, timeout=None, identifier=None
+    ):
+        with log_errors():
+            if not self._semaphore_exists(name):
+                return
```

Comment:
1. Acquire should return a boolean. The boolean is well defined and it is clear what's going on.
2. I would argue that an acquire for a not yet create semaphore should raise

Reply:
Would you also argue that it should raise for a semaphore that has already been closed?

Reply:
Without knowing too much about the situation here, I would default to whatever the behavior is for the normal Python Semaphore

Reply:
1. The python semaphore always returns bool
2. The "does not exist" does not exist in the python standard lib (since it's not distributed) so there are no references. I would argue this should raise in that scenario. 

this scenario would happen if

```
# client 1
semaphore = Semaphore(name="my-resource")

# client 2
semaphore = Semaphore(name="my-resource")

# client 1
semaphore.close()

# client 2
semaphore.acquire()
ValueError("Semaphore already closed")
```

as pointed out in the above [thread](https://github.com/dask/distributed/pull/3573/files/17ba039ff95dc84cb054e1d2872172bc1c3950ca#r396113995)

I believe this scenario should properly raise and we should make sure that Semaphore close should only be called by the process which is responsible for scheduling, e.g.

```
sem = Semaphore()
graph = my_dask_comp(sem=sem)
graph.compute()
sem.close()
```